### PR TITLE
Simplify InlineBuffer implementation

### PR DESCRIPTION
**Pull request by lopopolo**
This change was motivated by simplifying the memory layout of
`InlineBuffer` but ended up also reworking some of the core buffer
manipulation algorithms.

Turning `InlineBuffer` into an enum simplified handling the different
states which made it more clear which specialized APIs on `Vec` and
`ArrayVec` to make use of.

The switch to an enum and structuring code with match expressions
revealed some missing branches.

The resulting code makes heavy use of the `extend` family of APIs. There
is no more unsafe code for manipulating the `ArrayVec` outside of the
already unsafe `InlineBuffer::set_len`.

This commit fixes an off by one bug when detecting whether a slice can
be stored in the inline variant.

The new code makes more consistent use of `copied` iterators for
operating on slices.

Initial testing shows no significant performance changes when running
the spec suite, but the suite is dominated by `Regexp` performance. The
gains in correctness and simplicity afforded by this change make it the
right path forward.

Closes GH-414.

### CODE REVIEW

File: artichoke-backend/src/extn/core/array/inline_buffer.rs [hunk 1]
```diff
@@ -430,112 +413,122 @@ impl InlineBuffer {
...
-            for idx in 0..buflen {
-                ary.inline[idx] = unsafe { *buffer.get_unchecked(idx) };
-            }
-            for idx in buflen..start {
-                ary.inline[idx] = nil;
+    if start < INLINE_CAPACITY {
+        match ary {
+            InlineBuffer::Dynamic(buffer) => {
+                let mut inline = ArrayVec::new();
+                inline.extend(buffer.as_slice().iter().copied());
```
Comment (sjung-stripe) [inline artichoke-backend/src/extn/core/array/inline_buffer.rs]: could drain here

File: artichoke-backend/src/extn/core/array/inline_buffer.rs [hunk 2]
```diff
@@ -430,112 +413,122 @@ impl InlineBuffer {
...
+            InlineBuffer::Dynamic(ref mut buffer) => {
+                buffer.reserve(start + 1 - buflen);
+                for _ in buflen..start {
+                    buffer.push(nil);
+                }
+                buffer.push(elem.inner());
+            }
+            InlineBuffer::Inline(buffer) => {
+                let mut dynamic = vec![nil; start + 1];
+                dynamic[0..buflen].copy_from_slice(buffer.as_slice());
```
Comment (sjung-stripe) [inline artichoke-backend/src/extn/core/array/inline_buffer.rs]: could drain here

File: artichoke-backend/src/extn/core/array/inline_buffer.rs [hunk 3]
```diff
@@ -306,29 +276,36 @@ impl InlineBuffer {
...
+                } else if index < buffer.capacity() {
+                    let nil = unsafe { sys::mrb_sys_nil_value() };
+                    for _ in buflen..index {
+                        buffer.push(nil);
+                    }
+                    buffer.push(elem.inner());
+                } else {
+                    let nil = unsafe { sys::mrb_sys_nil_value() };
+                    let mut dynamic = vec![nil; index + 1];
+                    dynamic[0..buffer.len()].copy_from_slice(buffer.as_slice());
```
Comment (sjung-stripe) [inline artichoke-backend/src/extn/core/array/inline_buffer.rs]: could `buffer.drain(..)` here?

File: artichoke-backend/src/extn/core/array/inline_buffer.rs [hunk 4]
```diff
@@ -345,18 +322,21 @@ impl InlineBuffer {
...
-            for _ in buflen..start {
-                buffer.push(nil);
+            match self {
+                Self::Dynamic(ref mut buffer) => {
+                    buffer.push(with.inner());
+                }
+                Self::Inline(ref mut buffer) => {
+                    let nil = unsafe { sys::mrb_sys_nil_value() };
+                    let mut dynamic = vec![nil; start + 1];
+                    dynamic[0..buffer.len()].copy_from_slice(buffer.as_slice());
```
Comment (sjung-stripe) [inline artichoke-backend/src/extn/core/array/inline_buffer.rs]: could drain here

**Comment by lopopolo** [comment]
Thanks for the review @sjung-stripe. Your review prompted me to take a closer look and there are more bugs in there.

I've started adding some integration tests for the edges and when the data structure transitions from dynamic to inline.

**Comment by lopopolo** [comment]
latest commits add a boatload of tests

**Comment by lopopolo** [comment]
The integration tests resulted in lots of bugfixes in `InlineBuffer` itself (mostly panics in drain) as well as some bugs in the `Array` container and arg parsing.

**Comment by lopopolo** [comment]
rerunning failed windows build

**Comment by lopopolo** [comment]
Rebase to bring in GH-417 and fix the windows build

**Pull request merged**